Type updates

mr-c opened the pull request:
Why: https://github.com/common-workflow-language/galaxy/issues/137
What: Improved type hints

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

mr-c committed:
improve typing of lib/galaxy/workflow/run_request.py

mr-c committed:
improve typing of lib/galaxy/workflow/run.py

lib/galaxy/workflow/run.py:333: error: Item "DatasetCollection" of "Union[HistoryDatasetCollectionAssociation, DatasetCollection]" has no attribute "collection"

mr-c committed:
improve typing of lib/galaxy/workflow/modules.py

mr-c committed:
improve typing of lib/galaxy/util/compression_utils.py

mr-c committed:
improve typing of lib/galaxy/tools/wrappers.py

mr-c commented on setup.cfg:
```diff
@@ -218,7 +218,18 @@ check_untyped_defs = False
+disallow_untyped_decorators = True
+no_implicit_optional = True
+warn_unused_ignores = True
+warn_return_any = True
```
If `warn_return_any = True` is specified under `[mypy-galaxy.tools.wrappers]` then `tox -e test_galaxy_packages` fails

https://app.circleci.com/pipelines/github/galaxyproject/galaxy/19036/workflows/9cc3867a-c0f9-4d2e-8876-9a49ed4e54a6/jobs/156058?invite=true#step-104-2353

```
mypy galaxy tests
galaxy/tools/wrappers.py:262: error: Returning Any from function declared to return "Iterator[Any]"
galaxy/tools/wrappers.py:265: error: Returning Any from function declared to return "bool"
galaxy/tools/wrappers.py:346: error: Returning Any from function declared to return "List[Tuple[str, str]]"
galaxy/tools/wrappers.py:365: error: Returning Any from function declared to return "bool"
galaxy/tools/wrappers.py:573: error: Returning Any from function declared to return "List[List[Any]]"
```

but `tox -e py36-mypy` passes

davelopez commented on lib/galaxy/datatypes/binary.py:
```diff
@@ -281,8 +285,11 @@ def matches_any(self, target_datatypes):
         # CompressedArchive() in the following check.
-        return self.uncompressed_datatype_instance.matches_any(uncompressed_target_datatypes) or \
-            CompressedArchive().matches_any(compressed_target_datatypes)
+        if not hasattr(self, "uncompressed_datatype_instance"):
+            raise Exception("Missing 'uncompressed_datatype_instance' attribute.")
+        else:
```
Just out of curiosity (I'm still a Python initiate), is this necessary?
I would expect an AttributeError here without the condition, is it just to make the exception message more specific?

mr-c replied:
This is to satisfy the type checker. Nicer than an `assert`, in my opinion. Happy to change if that is preferred

davelopez replied:
Oh, I see, I don't have a strong opinion on whether to use assert or a full if/else in this regard, but maybe @jmchilton or others have one?

mr-c replied:
The other option would be a refactoring to ensure that `uncompressed_datatype_instance` is always present

davelopez replied:
That sounds even better :)

mr-c replied:
I lack the domain expertise to do a refactoring to ensure that `uncompressed_datatype_instance` is always present, so I would still request a decision on `assert` vs. `raise Exception` , so as to not block this PR :-)

mvdbeek replied:
I had suggested an `assert` in #12884, but I know @nsoranzo prefers raising an Exception. So that's a 2 to 1 vote, the Exception's have it.

davelopez commented on lib/galaxy/tools/wrappers.py:
```diff
@@ -3,8 +3,13 @@
 from galaxy import exceptions
+from galaxy.datatypes.registry import Registry
+from galaxy.jobs import ComputeEnvironment
+from galaxy.model import DatasetCollection, DatasetCollectionElement, DatasetCollectionInstance, DatasetInstance, HasTags, HistoryDatasetCollectionAssociation
```
I think this should be a multiline import. Same for `from typing import ...` above and other large imports.

mr-c replied:
Thanks. I just force-pushed new commits with better formatting (courtesy of `darker`)

davelopez commented on lib/galaxy/tools/wrappers.py:
```diff
@@ -385,46 +403,52 @@ def __getattr__(self, key):
     @staticmethod
-    def to_dataset_instances(dataset_instance_sources):
-        dataset_instances = []
+    def to_dataset_instances(dataset_instance_sources: Any) -> Sequence[Union[None, DatasetInstance]]:
```
```suggestion
    def to_dataset_instances(dataset_instance_sources: Any) -> List[Optional[DatasetInstance]]:
```

davelopez commented on lib/galaxy/workflow/modules.py:
```diff
@@ -830,42 +836,42 @@ def get_inputs(self):
-                input_default_value = TextToolParameter(None, default_source)
+                    text_default = ""
+                default_source["value"] = text_default
+                input_default_value: Union[TextToolParameter, IntegerToolParameter, FloatToolParameter, BooleanToolParameter, ColorToolParameter] = TextToolParameter(None, default_source)
```
Maybe just for legibility move the variable and type declaration right before the if?

mr-c replied:
I can't, as that would require an assignment to either `None` (which would break the type) or a default value (which may not be appropriate here, I don't know the code well enough to judge). However, the code is reformatted now, so maybe that is good enough?

davelopez reviewed (commented):
Thank you for all the type annotations! I learned a couple of things from it :)

mvdbeek reviewed (approved):
Very cool!

mr-c commented on lib/galaxy/datatypes/registry.py:
```diff
@@ -6,11 +6,12 @@
 import yaml
 
 import galaxy.util
+from galaxy.model import DatasetInstance
```
This could be moved under `if TYPE_CHECKING:` to reduce the chance of a future circular import issue

mr-c commented on lib/galaxy/tools/wrappers.py:
```diff
@@ -3,8 +3,33 @@
+    HasTags,
+    HistoryDatasetCollectionAssociation,
+)
+from galaxy.model.metadata import MetadataCollection
```
Likewise, these could also be moved under `if TYPE_CHECKING:`  to reduce the chances of cyclic imports in the future

jmchilton commented on lib/galaxy/workflow/run.py:
```diff
@@ -327,7 +336,7 @@ def replacement_for_connection(self, connection, is_data=True):
                     raise modules.CancelWorkflowEvaluation()
             else:
-                if not replacement.collection.populated:
+                if not replacement.collection.populated:  # type: ignore
```
I feel like an explicit cast() would be better than ``type: ignore``.

nsoranzo replied:
I think there is a real issue here, `model.DatasetCollection` doesn't have a `collection` attribute.

nsoranzo replied:
I pushed a commit to fix this (vetted by Marius).

nsoranzo committed:
Fix `WorkflowProgress.replacement_for_connection()` typing

The code wouldn't currently work if `replacement` is a `DatasetCollection`
because it tries to access the `collection` attribute which is
available only for `HistoryDatasetCollectionAssociation`.

nsoranzo commented on lib/galaxy/workflow/modules.py:
```diff
@@ -910,7 +926,9 @@ def get_inputs(self):
 
             if param_type == "text":
-                restrict_how_source = dict(name="how", label="Restrict Text Values?", type="select")
+                restrict_how_source: Union[
+                    Dict[str, Union[str, List[Dict[str, Union[str, bool]]]]]
+                ] = dict(name="how", label="Restrict Text Values?", type="select")
```
@mr-c  I've got Pylance in my editor complaining about this line:
```
Union requires two or more type arguments
```

Can I replace this with:
```python
                restrict_how_source: Dict[
                    str, Union[str, List[Dict[str, Union[str, bool]]]]
                ] = dict(name="how", label="Restrict Text Values?", type="select")
```

mr-c replied:
Fixed in https://github.com/galaxyproject/galaxy/pull/12899/commits/a467a7654e21663e7d7333891470c03d966a8db8
